ScrollViewer add property to bubble up the scroll event if not handled by nested ScrollViewer

timunie opened the pull request:
## What does the pull request do?
Add an AttachedProperty which can be used to bubble up the scrolling if a nested ScrollViewer didn't handle the scroll. 


## What is the current behavior?
The ScrollViewer eats up the PointerWheel event always, and doesn't respect if the scroll happened or not


## What is the updated/expected behavior with this PR?
Should fix issue #6869


## How was the solution implemented (if it's not obvious)?
- Added an attached property to the ScrollViewer
- Check if the scroll was handled or not and if not, bubble the event up if requested


## Checklist

- [ ] Added unit tests (if possible)?
- [X] Added XML documentation to any related classes?
- [x] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation ► I will need to do this, as it will be a new feature. But first the name and the interface should be approved. 
- [x] Find a way to respect the new property in DataGrid as well.
- [x] Remove Sandbox App

## Breaking changes
If the property should be `true` by defaulft, it is a breaking change. 

## Obsoletions / Deprecations
<!--- Obsolete and Deprecated attributes on APIs MUST only be included when discussed with Core team. @grokys, @kekekeks & @danwalmsley -->

## Fixed issues
Fixes #6869

timunie committed:
Implement `BubbleUpScrollOnEndReachedProperty`

- Add property as AttachedProperty, so it can be opt-in or opt-out
- Only bubble up the scroll event if the end is reached
- Make use of this Property in:
  o ListBox
  o TextBox
  o TreeView

timunie commented:
To see it in action please run the Sandbox App provided with this draft PR

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0017617-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json)

timunie committed:
use `ScrollViewer.GetBubbleUpScrollOnEndReached` in `DataGrid`

DataGrid handles the scrolling by it's own and not via ScrollViewer, so we need handle it there as well.

timunie commented:
> To see it in action please run the Sandbox App provided with this draft PR

This is not valid anymore. Thanks to @danwalmsley I can directly test the new feature in my sample. For testing please use this sample: https://github.com/timunie/Tims.Avalonia.Samples/tree/main/src/ScrollingIssueSample

Happy coding 
Tim

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0017709-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

maxkatz6 commented:
I was thinking about naming, and was pretty sure there should be something in other frameworks.

After looking at "brand-new" scrollviewer from WinUI, I found UWP actually has similar property:
https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.scrollviewer.ishorizontalscrollchainingenabled?view=winrt-22000#windows-ui-xaml-controls-scrollviewer-ishorizontalscrollchainingenabled
https://github.com/micahl/microsoft-ui-xaml-specs/blob/scrollviewer/active/ScrollViewer/ScrollViewer.md

Unfortunately one confusing part, UWP has API named IsHorizontalScrollChainingEnabled and they are going to rename it to HorizontalScrollChainingMode, though it wasn't confirmed and seems wasn't publicly discussed about this specific property (same for Vertical one).  I will try to check if this property actually does the same as BubbleUpScrollOnEndReachedProperty. And if yes, we should consider renaming it to either IsHorizontalScrollChainingEnabled or HorizontalScrollChainingMode (which is more flexibly enum).

timunie commented:
Hi @maxkatz6 

thank you for the valueable feedback.

I fully agree that this name is much more easy to understand. Also an enum would be fine, but tbh I don't know which other states then `enabled` or `disabled` are useful? 

Some considerations: 
-  You referenced the `IsHorizontalScrollChainingEnabled`. I think in Avalonia there is a `Vector` which handles both directions in one event, so `IsScrollChainingEnabled` may be better? 
- Also most Apps accept `[SHIFT]+[ScrollWheel]` for horizontal scroll direction. Maybe also this can be added in this PR or better in a seperate PR?

Happy coding 
Tim

maxkatz6 commented:
> Also most Apps accept [SHIFT]+[ScrollWheel] for horizontal scroll direction. Maybe also this can be added in this PR or better in a seperate PR?

Separated PR. Also it works already with macOS, I guess with native support. 


About IsHorizontalScrollChainingEnabled, it seems to work only with touch...for some reason. Weird behavior.

timunie commented:
> About IsHorizontalScrollChainingEnabled, it seems to work only with touch in uwp...for some reason. Weird behavior.

what does this mean for this PR? How should I proceed?

grokys commented:
To be honest I'm surprised this doesn't work by default - the fact that it doesn't work in WPF always really annoyed me and I intended to fix that in Avalonia when I wrote our `ScrollViewer`. So I'd say that the fact it doesn't currently work is a bug not an omission.

Given that, should we just fix it and make it chain scrolling by default? I'm not sure what the use-cases are for the property which enables/disables this behavior in UWP but I think in the (rare?) case where chaining is not wanted, it could be disabled in code.

maxkatz6 commented:
@grokys no, and it can't be enabled due IsHorizontalScrollChainingEnabled and IsVerticalScrollChainingEnabled working only with touch.
Let's add these two properties and allow them for all input types. CC @timunie 

Just for information, there also a CSS property https://developer.mozilla.org/en-US/docs/Web/CSS/overscroll-behavior

timunie commented:
@maxkatz6 sorry to bother you again. I also think a property is a good idea. But i am confused how to separate horizontal and vertical direction, as in Avalonia we use a Vector instead. So I think one property only would make more sense?!

maxkatz6 commented:
@timunie yeah, you are right. We can't handle event only partially... Let's go with "IsScrollChainingEnabled", just a name change.

grokys commented:
Single property 👍 from me, also I think it should default to `true` as it's the behavior that makes the most sense in the general case (assuming this change won't be backported to 0.11, which I don't think it will as we're in "bugfix only mode" there now).

Commit:
Rename BubbleUpScrollOnEndReached to IsScrollChainingEnabled

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0018247-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

maxkatz6 commented on src/Avalonia.Controls/Presenters/ScrollContentPresenter.cs:
```diff
@@ -60,6 +60,14 @@ public class ScrollContentPresenter : ContentPresenter, IPresenter, IScrollable,
+        /// Defines the <see cref="IsScrollChainingEnabled"/> property.
+        /// </summary>
+        public static readonly StyledProperty<bool> IsScrollChainingEnabledProperty =
+            AvaloniaProperty.Register<ScrollContentPresenter, bool>(
+                nameof(IsScrollChainingEnabled), 
+                defaultValue: true);
```
Can you try this change? I think it should work as well.
Because in current implementation we basically have two property registrations with the same name.

```suggestion
            ScrollViewer.IsScrollChainingEnabledProperty.AddOwner<ScrollContentPresenter>();
```

maxkatz6 replied:
Everything else looks good

timunie replied:
Updated

Commit:
Use AddOwner for ScrollContentPresenter

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0018293-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo commented:
This topic in UWP is a bit more complex than what is discussed here and I think it's useful to keep in mind for this discussion. UWP has [ManipulationMode](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.uielement.manipulationmode?view=winrt-22000) and the system can manage ScrollViewer specially. 

Assume there are two ScrollViewer nested within each other (there are stations this happens that make sense). If scroll speed is slow, the child ScrollViewer receives the scroll and moves. However, if the scroll speed is very fast the system automatically breaks out of the child and the parent manages the scroll instead.
